token.go

```diff
@@ -688,30 +692,64 @@ func processSingleResponse(sess *tdsSession, ch chan tokenStruct, outs outputs)
 			ch <- order
 		case tokenDoneInProc:
 			done := parseDoneInProc(sess.buf)
+
+			ch <- done
 			if sess.logFlags&logRows != 0 && done.Status&doneCount != 0 {
 				sess.log.Printf("(%d row(s) affected)\n", done.RowCount)
+
+				if outs.msgq != nil {
+					sqlexp.ReturnMessageEnqueue(ctx, outs.msgq, sqlexp.MsgRowsAffected{Count: int64(done.RowCount)})
```

Comment:
What should we do if ReturnMessageEnqueue fails? Simply returning the error doesn't seem helpful as it won't be any kind of error the caller can reasonable handle, is it? We definitely wouldn't want to have data/sql retry it.
Should we just panic?

Reply:
I looked at why it would return an error and it seems like overkill for it return ctx.Err() because the app is going to get that error back from Message() anyway.

```go
// Message is called by clients after Query to dequeue messages.
func (m *ReturnMessage) Message(ctx context.Context) RawMessage {
	select {
	case <-ctx.Done():
		return MsgNextResultSet{}
	case raw := <-m.queue:
		return raw
	}
}

// ReturnMessageEnqueue is called by the driver to enqueue the driver.
// Drivers should not call this until after it returns from Query.
func ReturnMessageEnqueue(ctx context.Context, m *ReturnMessage, raw RawMessage) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case m.queue <- raw:
		return nil
	}
}
```

Reply:
let me write a test where the context times out and doesn't return any results. I am not sure what  will happen :-)

Reply:
This is the basis of a test I think:
```go
ctx := context.WithTimeout(context.Background(),  5000*time.Millisecond)
	retmsg := &sqlexp.ReturnMessage{}
	r, err := conn.QueryContext(ctx, `waitfor delay '00:00:15'; select 100`, retmsg)
	
```
With my current implementation the client won't ever see a `sqlexp.MsgNext` because it waits until it gets columns to send it.  I think I can just post MsgNext as the first thing inside processSingleResponse

Reply:
then again, it seems the core data/sql code handles this already. This test passes if I don't do anything to handle errors from ReturnMessageEnqueue. The only message in the log is `MsgNextResultSet`

```go
func TestTimeoutWithNoResults(t *testing.T) {
	conn := open(t)
	defer conn.Close()
	latency, _ := getLatency(t)
	ctx, _ := context.WithTimeout(context.Background(), latency+5000*time.Millisecond)
	retmsg := &sqlexp.ReturnMessage{}
	r, err := conn.QueryContext(ctx, `waitfor delay '00:00:15'; select 100`, retmsg)
	if err != nil {
		t.Fatal(err.Error())
	}
	defer r.Close()
	active := true
	for active {
		msg := retmsg.Message(ctx)
		t.Logf("Got a message: %s", reflect.TypeOf(msg))
		switch m := msg.(type) {
		case sqlexp.MsgNextResultSet:
			active = r.NextResultSet()
			if active {
				t.Fatal("NextResultSet returned true")
			}
		case sqlexp.MsgNext:
			if r.Next() {
				t.Fatal("Got a successful Next even though the query should have timed out")
			}
		case sqlexp.MsgRowsAffected:
			t.Fatalf("Got a MsgRowsAffected %d", m.Count)
		}
	}
	if r.Err() != context.DeadlineExceeded {
		t.Fatalf("Unexpected error: %v", r.Err())
	}

}
```